### PR TITLE
Issue 20/add event member model

### LINKED ISSUE #20: [Ticket] Task 1 - EventMember Model & Migration
**Goal:** Establish the EventMember domain model and database structure so that each event registration is represented as a EventMember record.

**Pre‑requisites:**

* User and event models are already created with their basic relationships (an event belongs to an organizer).
* Database environment and migrations are set up.

**Steps:**

1. Generate a `EventMember` model with attributes:

   * `event_id` (foreign key to events),
   * `user_id` (foreign key to users),
   * `ticket_qr_code` (string, not null),
   * `rating` (smallint, optional; check 1 ≤ rating ≤ 5),
   * `comment` (text, optional)

2. Add database migrations:

   * Create the `EventMembers` table with the fields above.
   * Add indexes on `event_id`, `user_id`.
   * Add a unique index on `ticket_qr_code` to ensure no two tickets share the same code.
3. Define model associations: `belongs_to :event` and `belongs_to :user`.
4. Add basic validations: presence of `event_id` and `user_id`.
5. Seed the database with a few sample tickets to test relationships.
6. Write RSpec tests for model.

### PR DESCRIPTION
**Pull request by SomeTyraredGuy**

### CODE REVIEW

File: api/app/models/event_member.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class EventMember < ApplicationRecord
+  belongs_to :event
+  belongs_to :user
+
+  validates :user, presence: true
+  validates :event, presence: true
+  validates :ticket_qr_code, presence: true, uniqueness: true, length: { is: 10 }
```
Comment (mehalyna) [inline api/app/models/event_member.rb]: Validation for ticket_qr_code enforces a fixed length of 10, but generation logic may not guarantee uniqueness if input space is small. Consider adding a more robust uniqueness check or length validation logic.

File: api/db/seeds/event_members.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+def create_event_member!(event, user, rating: nil, comment: nil, qr_tag: nil)
+  qr = Digest::SHA256.hexdigest("#{user.id}:#{event.id}:#{qr_tag}")[0, 10]
```
Comment (mehalyna) [inline api/db/seeds/event_members.rb]: Digest::SHA256 for QR code generation may expose predictable user/event IDs. Consider using SecureRandom.hex for less predictable codes.
```
qr = SecureRandom.hex(5) # Generates a random 10-char string
```

File: api/db/seeds/event_members.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  EventMember.find_or_create_by!(ticket_qr_code: qr) do |em|
+    em.event = event
+    em.user = user
+    em.rating = rating
+    em.comment = comment
+  end
+end
+
+users = Role.for(:user).users.order(:id).to_a
+events = Event.order(:id).to_a
```
Comment (mehalyna) [inline api/db/seeds/event_members.rb:13]: Using all users and all events in seed may create excessive associations, slowing seed execution. Limit the number of associations or use sampling for large datasets.
```
users.sample(10).each_with_index do |user, i|
  # rest of seed logic
end
```

File: api/db/seeds/event_members.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+rated_event = events.first
+
+users.each_with_index do |user, i|
+  rand(1..3).times do |j|
+    create_event_member!(
+      events[(i + j) % events.size],
+      user,
+      qr_tag: "base_#{j}"
+    )
+  end
```
Comment (mehalyna) [inline api/db/seeds/event_members.rb]: The random number of event associations per user may result in uneven test coverage. Define explicit logic for event-member distribution in seed data.
```
users.each_with_index do |user, i|
  events.sample(2).each do |event|
    create_event_member!(event, user, qr_tag: "base_#{i}")
  end
end
```

File: api/db/seeds/event_members.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  create_event_member!(
+    rated_event,
+    user,
+    rating: rand(1..5),
+    comment: Faker::Lorem.sentence,
+    qr_tag: "rated_#{i}"
+  )
+end
```
Comment (mehalyna) [inline api/db/seeds/event_members.rb:28]: Using rand(1..5) for rating may not cover nil/blank cases. Add seed cases for EventMembers with nil ratings to test validation edge cases.
```
create_event_member!(
  rated_event,
  user,
  rating: [nil, rand(1..5)].sample,
  comment: Faker::Lorem.sentence,
  qr_tag: "rated_#{i}"
)
```

File: api/db/seeds/users.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+if ENV["SEED_RESET"] == "true"
+  UserRole.delete_all
+  User.delete_all
+end
+
+def create_user!(role, index)
+  user = User.find_or_create_by!(email: "#{role}_#{index}@example.com") do |u|
+      u.password = "password"
+      u.name = Faker::Name.name
+    end
```
Comment (mehalyna) [inline api/db/seeds/users.rb]: User email generation does not guarantee uniqueness if Role::NAMES contains duplicate values. Ensure email uniqueness by including role key in email format.
```
user = User.find_or_create_by!(email: "#{role}_#{index}_#{role_key}@example.com") do |u|
  # ...
end
```

File: api/db/seeds/users.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+  user.add_role!(role)
```
Comment (mehalyna) [inline api/db/seeds/users.rb]: add_role! is called for every user without checking existing roles, possibly causing unnecessary DB writes. Check and set user role only if not present.
```
user.add_role!(role) unless user.roles.include?(role)
```

File: api/db/seeds/events.rb [hunk 8]
```diff
@@ -1,4 +1,5 @@
 if ENV["SEED_RESET"] == "true"
+  EventMember.delete_all
```
Comment (mehalyna) [inline api/db/seeds/events.rb:2]: Deleting all EventMembers in SEED_RESET may break referential integrity if dependent events are deleted later. Wrap deletions in transaction and order for referential safety.
```
ActiveRecord::Base.transaction do
  EventMember.delete_all
  Event.delete_all
end
```

File: api/spec/models/event_member_spec.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    subject do
+      described_class.new(
+        event: Event.new(
+          title: "Test",
+          location: "Kyiv",
+          start_date: 2.days.from_now,
+          finish_date: 3.days.from_now,
+          participant_capacity: 100
+        ),
+        user: User.new(name: "J", email: "j@example.com", password: "password123"),
```
Comment (mehalyna) [inline api/spec/models/event_member_spec.rb]: Creating User and Event objects inline may not persist them, risking validation failures. Use FactoryBot or persist objects before testing validations.
```
event = FactoryBot.create(:event)
user = FactoryBot.create(:user)
subject { described_class.new(event: event, user: user, ticket_qr_code: "ABCDEF1234") }
```

File: api/db/seeds/event_members.rb (hunk 3 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/event_members.rb:13]: Done

File: api/db/seeds/event_members.rb (hunk 4 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/event_members.rb]: Done

File: api/db/seeds/users.rb (hunk 6 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/users.rb]: Done

File: api/db/seeds/events.rb (hunk 8 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/events.rb:2]: Done

File: api/db/seeds/event_members.rb (hunk 5 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/event_members.rb:28]: A nil rating is undesirable when a comment is present. This comment relates to a review. Added the `require_rating_for_comment` validation to prevent this issue

File: api/db/seeds/users.rb (hunk 7 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/users.rb]: add_role! from User model already guards against duplicates

```
  def add_role!(key)
    role = Role.for(key)
    roles << role unless roles.exists?(id: role.id)
  end
 ```

File: api/db/seeds/event_members.rb (hunk 2 above)
Comment (SomeTyraredGuy) [reply api/db/seeds/event_members.rb]: The idea was to make seeds idempotent and save to re-run. With SecureRandom.hex qr codes would be different every time and every re-run would create new `EventMember` records (its only unique field and it is used in find_or_create_by!). Should I still use SecureRandom.hex?

**Review by lhalam — COMMENTED** [review]
update

**Review by a2teak — DISMISSED** [review]
Looks good to me

File: api/spec/models/event_member_spec.rb (hunk 9 above)
Comment (SomeTyraredGuy) [reply api/spec/models/event_member_spec.rb]: Done

File: api/app/models/event_member.rb (hunk 1 above)
Comment (SomeTyraredGuy) [reply api/app/models/event_member.rb]: Changed validation to length: { minimum: 10, maximum: 18 }. Created const QR_CODE_LENGTH to be used in creation and changed it to 12.

File: api/db/seeds/events.rb [hunk 10]
```diff
@@ -1,5 +1,8 @@
 if ENV["SEED_RESET"] == "true"
-  Event.delete_all
+  ActiveRecord::Base.transaction do
+    EventMember.delete_all
+    Event.delete_all
+  end
```
[the remark this answers is not in the document]
Comment (SomeTyraredGuy) [reply api/db/seeds/events.rb]: Done

File: api/app/models/event_member.rb (hunk 1 above)
Comment (SomeTyraredGuy) [reply api/app/models/event_member.rb]: Length changed to 36, so it matches UUID format

**Review by mehalyna — APPROVED** [review]

**Review by lhalam — APPROVED** [review]

**Review by a2teak — APPROVED** [review]
lgtm

**Review by a2teak — APPROVED** [review]
lgtm
